scripts/util/doc-builder.js

```diff
@@ -0,0 +1,240 @@
+    return yuidoc.generate({paths: paths, themeDir: this.options.themeDir}).then(() => {
+      console.info('info:', 'Done generating API docs');
+    });
+
+  }
+
+  commitAPIDocs() {
+    var apiTempDir = path.join('.', this.options.versionsDirName);
+
+    console.info('\ninfo:', 'Commiting API Docs');
+    if (fs.existsSync(apiTempDir)) {
+      fsExtra.removeSync(apiTempDir);
+    }
+
+    console.info('info:', 'Copying', this.options.apiDirPath, 'to root directory');
+    fsExtra.copySync(this.options.apiDirPath, apiTempDir);
+
+    return NodeGit.Repository.open('.').then(repo => {
+
+      console.info('info:', 'Getting repo\'s current staging area');
+      return repo.index().then(index => {
+        console.info('info:', 'Getting', this.options.docsBranchName, '\'s latest commit');
+        return repo.getBranchCommit(this.options.docsBranchName).then(docsBranchCommit => {
+          console.info('info:', 'Getting the commit\'s tree');
+          return docsBranchCommit.getTree().then(tree => {
+            console.info('info:', 'Loading the tree into the staging area');
+            return index.readTree(tree);
+          }).then(() => {
+            console.info('info:', 'Staging \'', this.options.versionsDirName, '\' to \'', this.options.docsBranchName, '\'');
+            return index.addAll(`${this.options.versionsDirName}`, NodeGit.Index.ADD_OPTION.ADD_FORCE);
+          }).then(() => {
+            return repo.getReference(this.options.docsBranchName);
+          }).then(reference => {
+            var docsBranchResolvedName = reference.name();
+            console.info('info:', 'Blew up', this.options.docsBranchName, 'into', docsBranchResolvedName);
+            console.info('info:', 'Getting the staging area tree id');
+            return index.writeTree().then(treeId => {
+              var dateObject = new Date();
+              var time = dateObject.getTime() / 1000;
+              var timeOffset = dateObject.getTimezoneOffset();
+              var author = NodeGit.Signature.create("Auto Docs",
+                "docs@shopify.com", time, timeOffset);
+              var committer = NodeGit.Signature.create("Auto Docs",
+                "docs@shopify.com", time, timeOffset);
+
+              console.info('info:', 'Committing changes to \'', docsBranchResolvedName, '\'');
+              return repo.createCommit(`${docsBranchResolvedName}`, author, committer, 'Docs auto updated during build process', treeId, [docsBranchCommit]); 
+            }).then(commitId => {
+              console.info('info:', 'New commit created:', commitId);
+              console.info('info:', 'Deleting', this.options.versionsDirName, 'from the root directory');
+              fsExtra.removeSync(path.join('.', this.options.versionsDirName));
+
+              console.info('info:', 'Done commiting API docs');
+            });
+          });
+        });
+      });
+    });
+  }
+
+  build() {
+    return this.ensureDocsBranchExists().then(() => {
```

Comment:
@yomexzo I know we sat down to discuss this in person but I'm adding this here as a note. The following could be done to the `build` function and then the other functions could be cleaned up in a similar method.

1. You could get and save the `repo` to the class scope since all subsequent functions request the repo everytime. This would eliminate `NodeGit.Repository.open('.')` calls down the line. Also it'll make it so that the promise trees remain at one depth.
2. The promise tree in the build function could be made "flat"/or at one depth (example below)
3. Some of the function names could be made more descriptive to allow for the promise chain to be more semantic. So you should be able to literally read the code as "getTheRepo __THEN__ ensureDocsBranchExists __THEN__ checkoutDocsBranch..." (example below)

Example:
```javascript
build() {
  NodeGit.Repository.open('.').then((repo) => {
    this.repo = repo;
  })
  .then(this.ensureDocsBranchExists)
  .then(this.checkoutDocsBranch)
  .then(this.getLatestSrcPaths)
  .then(this.generateAPIDocsIntoGeneratedPaths)
  .then(() => {
    if (this.options.rmSrcDir) {
      console.info('\ninfo:', 'Removing source directories');
      paths.forEach(item => {
        var srcPath = path.join(item, this.options.srcDirName);
        console.info('info:', 'Removing', srcPath);
        fsExtra.removeSync(srcPath);
      });
    }

    if (this.options.commitDocs) {
      return this.commitAPIDocs();
    }
  });
}
```

Also maybe the last anonymous function which gets called could be named just so it's all readable.